kernel/src/dynamic_process_loading.rs

```diff
@@ -0,0 +1,958 @@
+        self.client.set(client);
+    }
+
+    fn setup(&self, app_length: usize) -> Result<usize, ErrorCode> {
+        //TODO(?): Check if it is a newer version of an existing app.
+        // We can potentially flash the new app and load it before erasing the old one.
+        // What happens to the process though? Need to delete it from the process array and load it back in?
+
+        let mut new_app_data = NewApp {
+            // struct to hold some information about the new app
+            size: 0,
+            start_addr: 0,
+        };
+
+        let flash_start = self.flash.get().as_ptr() as usize; //start of the flash region
+
+        new_app_data.start_addr = self.flash.get().as_ptr() as usize;
+        new_app_data.size = app_length;
+
+        match self.find_next_available_address(&mut new_app_data) {
+            Ok(()) => {
+                let new_start_addr = new_app_data.start_addr;
+
+                let offset = new_start_addr - flash_start;
+
+                let new_process_flash = self
+                    .flash
+                    .get()
+                    .get(offset..offset + app_length)
+                    .ok_or(ErrorCode::FAIL)?;
+
+                self.new_process_flash.set(new_process_flash);
+
+                self.new_app_start_addr.set(new_app_data.start_addr);
+                self.new_app_length.set(new_app_data.size);
+
+                // reset the struct values for a new app
+                new_app_data.size = 0;
+                new_app_data.start_addr = 0;
+
+                Ok(app_length)
+            }
+            Err(_err) => Err(ErrorCode::FAIL),
+        }
+    }
+
+    fn write_app_data(
+        &self,
+        flag: bool,
+        buffer: &'static mut [u8],
+        offset: usize,
+        length: usize,
+        _processid: ProcessId,
+    ) -> Result<(), ErrorCode> {
+        if !flag {
+            // this flag checks if the result is from a command that can be executed now, or if the command has been tabled for later
+            self.current_user.set(NonvolatileUser::Client);
+            let res = self.enqueue_command(
+                false,
+                NonvolatileCommand::UserspaceWrite,
+                buffer,
+                offset,
+                length,
+            );
+            match res {
+                Ok(()) => Ok(()),
+                Err(e) => Err(e),
+            }
+        } else {
+            // unimplemented!();   //when there are pending commands (the capsule currently returns something but we don't want it to do anything here)
+            Err(ErrorCode::NOSUPPORT)
+        }
+    }
+
+    fn load(&self) -> Result<(), ErrorCode> {
+        let process_flash = self.new_process_flash.take().ok_or(ErrorCode::FAIL)?;
+        let remaining_memory = self.app_memory.take().ok_or(ErrorCode::FAIL)?;
+
+        // Get the first eight bytes of flash to check if there is another app.
```

Comment:
Another app?

Reply:
I forgot to fix this comment, I have it fixed but not pushed. It should be fixed the next time I push the code.